agrirouter-sdk-java-impl/src/main/java/com/dke/data/agrirouter/impl/messaging/encoding/EncodeMessageServiceImpl.java

```diff
@@ -29,26 +51,155 @@ public String encode(
+      PayloadParameters payloadParameters,
+      OnboardingResponse onboardingResponse) {
+    logMethodBegin(messageHeaderParameters, payloadParameters);
+
+    if (null == messageHeaderParameters
+        || null == payloadParameters
+        || null == onboardingResponse) {
+      throw new IllegalArgumentException("Parameters cannot be NULL");
+    }
+    messageHeaderParameters.validate();
+    payloadParameters.validate();
+
+    if (null != messageHeaderParameters.getTechnicalMessageType()
+        && messageHeaderParameters.getTechnicalMessageType().getNeedsChunking()) {
+      if (payloadParameters.shouldBeChunked()) {
+        getNativeLogger()
+            .debug(
+                "The message should be chunked, current size of the payload ({}) is above the limitation.",
+                payloadParameters.getValue().toStringUtf8().length());
+        String wholeMessage = payloadParameters.getValue().toStringUtf8();
+        final List<String> messageChunks =
+            Splitter.fixedLength(payloadParameters.maxLengthForMessages())
+                .splitToList(wholeMessage);
+        List<MessageParameterTuple> tuples = new ArrayList<>();
+        AtomicInteger chunkNr = new AtomicInteger(1);
+        final String chunkContextId = ChunkContextIdService.generateChunkContextId();
+        messageChunks.forEach(
+            chunk -> {
+              final String messageIdForChunk = MessageIdService.generateMessageId();
+              final long sequenceNumberForChunk =
+                  SequenceNumberService.generateSequenceNumberForEndpoint(onboardingResponse);
+
+              final MessageHeaderParameters header = new MessageHeaderParameters();
+              header.copy(messageHeaderParameters);
+              header.setApplicationMessageId(messageIdForChunk);
+              header.setApplicationMessageSeqNo(sequenceNumberForChunk);
+              Chunk.ChunkComponent.Builder chunkInfo = Chunk.ChunkComponent.newBuilder();
+              chunkInfo.setContextId(chunkContextId);
+              chunkInfo.setCurrent(chunkNr.get());
+              chunkInfo.setTotal(messageChunks.size());
+              chunkInfo.setTotalSize(wholeMessage.length());
+              header.setChunkInfo(chunkInfo.build());
+
+              final PayloadParameters payload = new PayloadParameters();
+              payload.copyFrom(payloadParameters);
+              payload.setValue(
+                  ByteString.copyFromUtf8(
+                      Base64.getEncoder().encodeToString(chunk.getBytes(StandardCharsets.UTF_8))));
+
+              tuples.add(new MessageParameterTuple(header, payload));
+
+              chunkNr.getAndIncrement();
+            });
+        return tuples;
+      } else {
+        getNativeLogger()
+            .debug(
+                "The message is not chunked since the current size of the payload ({}) is not above the limitation or the technical message type '{}' does not support chunking.",
```

Comment:
the part "or the technical message type '{}' does not support chunking" can be removed as - if we reach this point - it was checked that we support chunking but don't need it